Store arrays offsets for keyword fields natively with synthetic source

martijnvg opened the pull request:
The keyword doc values field gets an extra sorted doc values field, that encodes the order of how array values were specified at index time. This also captures duplicate values. This is stored in an offset to ordinal array that gets zigzag vint encoded into a sorted doc values field. 

For example, in case of the following string array for a keyword field: ["c", "b", "a", "c"].
Sorted set doc values: ["a", "b", "c"] with ordinals: 0, 1 and 2. The offset array will be: [2, 1, 0, 2]

Null values are also supported. For example  ["c", "b", null, "c"] results into sorted set doc values: ["b", "c"] with ordinals: 0 and 1. The offset array will be: [1, 0, -1, 1]

Empty arrays are also supported by encoding a zigzag vint array of zero elements.

Limitations:
* currently only doc values based array support for keyword field mapper.
* multi level leaf arrays are flattened. For example: [[b], [c]] -> [b, c]
* arrays are always synthesized as one type. In case of keyword field, [1, 2] gets synthesized as ["1", "2"].

These limitations can be addressed, but some require more complexity and or additional storage.

With this PR, keyword field array will no longer be stored in ignored source, but array offsets are kept track of in an adjacent sorted doc value field. This only applies if `index.mapping.synthetic_source_keep` is set to `arrays` (default for logsdb).

martijnvg committed:
Synthetic source doc values arrays encoding experiment

The keyword doc values field gets an extra binary doc values field, that encodes the order of how array values were specified at index time. This also captures duplicate values.

This is stored in an offset to ordinal array that gets vint encoded into the binary doc values field. The additional storage required for this will likely be minimized with #112416 (zstd compression for binary doc values)

In case of the following string array for a keyword field: ["c", "b", "a", "c"].
Sorted set doc values: ["a", "b", "c"] with ordinals: 0, 1 and 2. The offset array will be: [2, 1, 0, 2]

Limitations:
* only support for keyword field mapper.
* multi level leaf arrays are flattened. For example: [[b], [c]] -> [b, c]
* empty arrays ([]) are not recorded
* arrays are always synthesized as one type. In case of keyword field, [1, 2] gets synthesized as ["1", "2"].

These limitations can be addressed, but some require more complexity and or additional storage.

martijnvg committed:
spotless

martijnvg committed:
Merge remote-tracking branch 'es/main' into synthetic_source_encode_arrays

martijnvg committed:
iter

salvatore-campagna commented on server/src/main/java/org/elasticsearch/index/mapper/SortedSetDocValuesSyntheticFieldLoaderLayer.java:
```diff
@@ -133,6 +148,87 @@ public void write(XContentBuilder b) throws IOException {
+                    ords[i] = dv.nextOrd();
+                }
+
+                logger.info("ords=" + Arrays.toString(ords));
```
I guess you are using `info` just for debugging while in draft.

martijnvg replied:
Yes, this was for debugging purposes. This will be removed.

martijnvg commented on server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java:
```diff
@@ -856,4 +859,27 @@ public String currentName() throws IOException {
+    }
+
+    public void recordOffset(String fieldName, String value) {
+        int count = numValuesByField.compute(fieldName, (s, integer) -> integer == null ? 0 : ++integer);
```
`count` is the wrong name here. It is like the next offset to be used.

martijnvg committed:
Merge remote-tracking branch 'es/main' into synthetic_source_encode_arrays

martijnvg committed:
iter

martijnvg committed:
Merge remote-tracking branch 'es/main' into synthetic_source_encode_arrays

martijnvg committed:
spotless

martijnvg committed:
Merge remote-tracking branch 'es/main' into synthetic_source_encode_arrays

lkts commented on server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java:
```diff
@@ -906,17 +924,51 @@ private KeywordFieldMapper(
+        }
+    }
+
+    public void processOffsets(DocumentParserContext context) throws IOException {
```
Should this be called from `postParse`? It is possible that a field is indexed multiple times in one document with object arrays.

lkts replied:
I am surprised randomized tests don't complain about this.

martijnvg replied:
👍 I think postParse is a better place to invoke this process offsets logic.

lkts reviewed (commented):
This is a very powerful idea. Some thoughts:
* When we generalize this, we need to think how to fit it into existing code nicely (leave poor `DocumentParser` alone). I was thinking lately that maybe `DocumentParser` can produce events like "parsing array", "parsing object", "parsing value" and then we can subscribe to such events and do our thing. Didn't dive too deep into this though.
* I wonder if we can have a byte or two in the beginning of such encoding that can carry meta information. An example would be an "empty array" flag or "single element array".

martijnvg commented:
Thanks for taking a look @lkts!

>  I was thinking lately that maybe DocumentParser can produce events like "parsing array", "parsing object", "parsing value" and then we can subscribe to such events and do our thing. Didn't dive too deep into this though.

I think we already have something like this via the `FieldMapper#parsesArrayValue()` flag. I initially tried using this, because I don't like introducing more complexity in `DocumentParser`. However it didn't work out in all cases. I recall that tests using copy to failed. For some reason field that overwrite that method and return true, are not taken into account with copy to. Maybe we should make fields that chose to overwrite `FieldMapper#parsesArrayValue()` work  with copy_to correctly. 

> I wonder if we can have a byte or two in the beginning of such encoding that can carry meta information. An example would be an "empty array" flag or "single element array".

Good point, I think we can have an info byte where this kind of information can be encoded. I was thinking something similar earlier, but left this out of this draft PR in order to keep it simple for demonstration purposes.

martijnvg committed:
Merge remote-tracking branch 'es/main' into synthetic_source_encode_arrays

martijnvg committed:
parsesArrayValue approach

martijnvg committed:
fix multi fields

martijnvg committed:
iter

martijnvg committed:
Merge remote-tracking branch 'es/main' into synthetic_source_encode_arrays

martijnvg committed:
do not handle copy_to for now

martijnvg committed:
cleanup

martijnvg committed:
move ValueXContentParser

martijnvg committed:
adjust expected json element type

martijnvg commented:
@salvatore-campagna and @lkts I made a few changes and keyword field mapper now overwrites `parsesArrayValue()`, so that it can parse arrays. This allows minimizing changes in `DocumentParser`.

martijnvg committed:
Merge remote-tracking branch 'es/main' into synthetic_source_encode_arrays

martijnvg committed:
iter

martijnvg committed:
fixed mistake

martijnvg committed:
Merge remote-tracking branch 'es/main' into synthetic_source_encode_arrays

martijnvg committed:
Merge remote-tracking branch 'es/main' into synthetic_source_encode_arrays

martijnvg committed:
iter

martijnvg committed:
iter

martijnvg committed:
Merge remote-tracking branch 'es/main' into synthetic_source_encode_arrays

martijnvg committed:
Moved processing of offsets to DocumentParser instead of in fieldmapper, because otherwise the offset binary doc values field gets added multiple times if keyword has array and objects fields as parent (lucene validation fails in that case).

Also removed the logic that handles multi field logic and instead rely on fallback to handle if multi fields are configured.

And a test cleanup in synthetic source support.

Commit:
[CI] Auto commit changes from spotless

martijnvg committed:
* Added support for json null by encoding it as \0
* Added support for empty arrays, by storing offset array of size zero.
* removed dev logging / todos
* added more tests

martijnvg committed:
Encode offsets using zigzag encoding, so that -1 can be used to encode json null values.

martijnvg committed:
fixed codec issue if only NULLs gets indexed

martijnvg committed:
Update docs/changelog/113757.yaml

martijnvg commented:
I've updated this PR and made a few changes:

- Moved the processing of collected offsets from `KeywordFieldMapper` to `DocumentParser`.  The issue with having this logic in `KeywordFieldMapper` is that we can add a binary doc values field multiple times if a leaf field is part of a parent object array field.
- Updated the encoding for storing the offset array from vint array to zigzag encoded vint array. This allows to use negative values. The value -1 is used to encode json null values.
- Empty arrays are now supported.
- More tests

martijnvg committed:
Merge remote-tracking branch 'es/main' into synthetic_source_encode_arrays

elasticsearchmachine commented:
Pinging @elastic/es-storage-engine (Team:StorageEngine)

lkts commented on server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java:
```diff
@@ -867,14 +896,18 @@ public boolean hasNormalizer() {
     private final IndexAnalyzers indexAnalyzers;
     private final int ignoreAboveDefault;
     private final int ignoreAbove;
+    private final BinaryFieldMapper offsetsFieldMapper;
```
Do we actually need this? We only ever use `fullPath()` from it.

martijnvg replied:
Right, I think we can remove it. We can just keep track of offsetFieldName.

I think I wanted to share this field via `#iterator()`, so that this field is visible in other places like field caps, but I think this field should remain internal/hidden. Additionally `BinaryFieldMapper` uses a custom binary doc values field  that has no advantage for offsets.

martijnvg committed:
Merge remote-tracking branch 'es/main' into synthetic_source_encode_arrays

lkts commented on server/src/main/java/org/elasticsearch/index/mapper/SortedSetDocValuesSyntheticFieldLoaderLayer.java:
```diff
@@ -133,6 +149,98 @@ public void write(XContentBuilder b) throws IOException {
         }
     }
 
+    class OffsetDocValuesLoader implements DocValuesLoader, DocValuesFieldValues {
```
I think this would be better structured if this was a completely new layer (it looks like all existing functionality is overridden). Then if `offsetsFullPath` is not null you use the new layer, and if it is you continue to use `SortedSetDocValuesSyntheticFieldLoaderLayer`.

martijnvg replied:
Agreed, a new layer makes more sense and is cleaner.

lkts commented on server/src/test/java/org/elasticsearch/index/mapper/SyntheticSourceNativeArrayIntegrationTests.java:
```diff
@@ -0,0 +1,139 @@
+        verifySyntheticArray(arrayValues);
+    }
+
+    public void testSynthesizeArrayRandom() throws Exception {
```
Do we test with `ignore_above` anywhere?

martijnvg replied:
Added in `SyntheticSourceNativeArrayIntegrationTests#testSynthesizeArrayIgnoreAbove(...)`

martijnvg committed:
handle string arrays for keyword fields nested under object arrays correctly.
The offset optimization will not kick in and instead rely on ignored source.

martijnvg committed:
offsetsFieldMapper -> offsetsFieldName

martijnvg committed:
added ignore above test

martijnvg committed:
Move OffsetDocValuesLoader to its own layer implementation

martijnvg committed:
iter

martijnvg committed:
Merge remote-tracking branch 'es/main' into synthetic_source_encode_arrays

kkrik-es commented on server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java:
```diff
@@ -470,6 +483,35 @@ public Set<String> getCopyToFields() {
+
+    void maybeRecordEmptyArray(String field) {
+        offsetsPerField.computeIfAbsent(field, k -> new Offsets());
+    }
```
Nit: add a static class `FieldOffsets` to encapsulate these methods, along with the `Map<String, Offsets>`? It can then be unit-tested too.

martijnvg replied:
Good point. See: 01cd313546089f363a6ae9ba9c6c8c9ea16b6e23

kkrik-es commented on server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java:
```diff
@@ -899,11 +932,64 @@ public KeywordFieldType fieldType() {
+            }
+        } else {
+            super.parse(context);
+        }
```
Can we keep this logic in `DocumentParser`? It feels like leaking parsing implementations in mappers..

martijnvg replied:
I made the following change: 012ac7f0f3ef5d302a5343479f82a2124f322fb7
In order to make sure empty arrays and nested leaf arrays are covered some logic had to be added. Let me know what you think of this. The nice thing is that now adding offset to other field types should be easier.

The previous approach was more isolated, field types would be responsible for parsing arrays, which means not many changes to DocumentParser were needed. I think I like this PR with 012ac7f0f3ef5d302a5343479f82a2124f322fb7 more now.

kkrik-es replied:
Nit: `return (offsetsFieldName != null);`

kkrik-es commented on server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java:
```diff
@@ -899,11 +932,64 @@ public KeywordFieldType fieldType() {
+    private void parseArray(DocumentParserContext context) throws IOException {
+        XContentParser parser = context.parser();
+        while (true) {
+            XContentParser.Token token = parser.nextToken();
```
Ditto, this feels out of place here, belongs to document parsing.

kkrik-es commented on server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:
```diff
@@ -166,7 +166,7 @@ public boolean ignoreMalformed() {
      * and each value is passed to the mapper for parsing.
      */
-    public boolean parsesArrayValue() {
+    public boolean parsesArrayValue(DocumentParserContext context) {
```
Would it suffice to add another method in this class, indicating if the `FieldMapper` supports tracking leaf arrays? It seems like `DocumentParser` could use that to switch between the two parsing modes and track offsets when needed.

kkrik-es commented on server/src/main/java/org/elasticsearch/index/mapper/SortedSetWithOffsetsDocValuesSyntheticFieldLoaderLayer.java:
```diff
@@ -0,0 +1,170 @@
+
+    static final class ImmediateDocValuesLoader implements DocValuesLoader {
+        private final BinaryDocValues oDv;
+        private final SortedSetDocValues dv;
```
Nit: let's rename to `ordinalDocValues` and `docValues` for readability.

martijnvg committed:
rename

martijnvg committed:
move some offset logic from DocumentParserContext to new ArrayOffsetContext

martijnvg committed:
use correct full name for offset field

martijnvg committed:
Move offset parsing logic to DocumentParser

Commit:
[CI] Auto commit changes from spotless

kkrik-es commented on server/src/main/java/org/elasticsearch/index/mapper/ArrayOffsetContext.java:
```diff
@@ -0,0 +1,81 @@
+import java.util.Map;
+import java.util.TreeMap;
+
+public class ArrayOffsetContext {
```
FieldArrayContext? We may add more things than offsets, as needed.

kkrik-es commented on server/src/main/java/org/elasticsearch/index/mapper/ArrayOffsetContext.java:
```diff
@@ -0,0 +1,81 @@
+        offsetsPerField.computeIfAbsent(field, k -> new Offsets());
+    }
+
+    void processArrayOffsets(DocumentParserContext context) throws IOException {
```
Rename to `writeToDoc` or something? `process*` is too generic..

kkrik-es commented on server/src/main/java/org/elasticsearch/index/mapper/ArrayOffsetContext.java:
```diff
@@ -0,0 +1,81 @@
+        }
+    }
+
+    static class Offsets {
```
private?

kkrik-es commented on server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java:
```diff
@@ -891,17 +921,43 @@ private KeywordFieldMapper(
+            } else if (value == null) {
+                context.getOffSetContext().recordNull(offsetsFieldName);
+            }
+        }
```
Same.

martijnvg replied:
I think we need to keep null handling in the field mappers? Only if value is not indexed and is null (which depends on null_value configuration) then we record null, which we only when invoking `parseCreateField(...)`?

martijnvg committed:
Merge remote-tracking branch 'es/main' into synthetic_source_encode_arrays

martijnvg committed:
synthetic recovery source is enabled by default now (in snapshot builds)

martijnvg commented on server/src/main/java/org/elasticsearch/index/mapper/ArrayOffsetContext.java:
```diff
@@ -0,0 +1,81 @@
+    static class Offsets {
+
+        public int currentOffset;
+        public final Map<String, List<Integer>> valueToOffsets = new TreeMap<>();
```
This is needed otherwise we don't store the offsets in the right order. We need to iterate the unique values with offset in the order we insert (which is the same as the order in which we parse). I will add a comment.

Note that many tests added in this PR fail if you change this to `HashMap`.

martijnvg committed:
Small iter

martijnvg committed:
handle empty array logic in DocumentParser

kkrik-es commented on server/src/test/java/org/elasticsearch/index/mapper/KeywordFieldTypeTests.java:
```diff
@@ -244,7 +244,8 @@ public void testFetchSourceValue() throws IOException {
             Integer.MAX_VALUE,
-            IndexVersion.current()
+            IndexVersion.current(),
+            Mapper.SourceKeepMode.NONE
```
randomFrom?

kkrik-es commented on server/src/test/java/org/elasticsearch/index/mapper/OffsetDocValuesLoaderTests.java:
```diff
@@ -0,0 +1,113 @@
+    }
+
+    private void verifyOffsets(String source, String expectedSource) throws IOException {
+        var mapperService = createMapperService("""
```
Why not just call `MapperServiceTestCase.syntheticSource`?

martijnvg replied:
This is also kind of a direct unit test against `SortedSetWithOffsetsDocValuesSyntheticFieldLoaderLayer` and `MapperServiceTestCase.syntheticSource(...)` requires XContentBuilder, and here I assert directly with documents as strings.

kkrik-es commented on server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java:
```diff
@@ -376,9 +377,11 @@ private static void innerParseObject(DocumentParserContext context) throws IOExc
                     }
                     break;
                 case START_OBJECT:
+                    context.setImmediateXContentParent(token);
```
Move this in `parseObject` and `parseArray` to avoid duplication?

martijnvg committed:
Merge remote-tracking branch 'es/main' into synthetic_source_encode_arrays

martijnvg committed:
randomFrom

martijnvg committed:
reduce number of `setImmediateXContentParent(...)` invocations

martijnvg commented on server/src/test/java/org/elasticsearch/index/mapper/OffsetDocValuesLoaderTests.java:
```diff
@@ -0,0 +1,113 @@
+    protected Settings getIndexSettings() {
+        return Settings.builder()
+            .put("index.mapping.source.mode", "synthetic")
+            .put("index.mapping.synthetic_source_keep", "arrays")
```
Yes, if `index.mapping.synthetic_source_keep` is set to `none` then array offsets aren't stored. If set to `all` then we store arrays using ignored source.

We can maybe randomize between `index.mapping.source.mode` set to `synthetic` (with `index.mapping.synthetic_source_keep`) and `index.mode` set to logsdb?

kkrik-es commented on server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java:
```diff
@@ -729,13 +737,16 @@ private static void parseNonDynamicArray(
+            // elements from regular source loading that are then skipped from the synthesized array source.
+            // To prevent this, we track that parsing sub-context is within array scope.
+            context = context.maybeCloneForArray(mapper);
+        }
```
Move to the branch above, to avoid repeating the condition?

kkrik-es commented on server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java:
```diff
@@ -754,6 +765,14 @@ private static void parseNonDynamicArray(
             }
+            previousToken = token;
+        }
+        if (mapper != null
```
`context.canAddIgnoredField()` too

martijnvg replied:
done

kkrik-es commented on server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java:
```diff
@@ -611,6 +615,8 @@ private static void throwOnCreateDynamicNestedViaCopyTo(Mapper dynamicObjectMapp
     }
 
     private static void parseArray(DocumentParserContext context, String lastFieldName) throws IOException {
+        var prev = context.getImmediateXContentParent();
```
Consider adding a comment here. I was puzzled why we need this in object and array, but then thought about (a) leaf arrays within leaf arrays and (b) object arrays.

kkrik-es commented on server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java:
```diff
@@ -460,6 +487,33 @@ public boolean isCopyToDestinationField(String name) {
+        return fieldArrayContext;
+    }
+
+    private XContentParser.Token token;
```
Ditto, maybe add a comment explaining why we need to track this, and rename to `lastParsedToken`?

kkrik-es commented on server/src/main/java/org/elasticsearch/index/mapper/FieldArrayContext.java:
```diff
@@ -0,0 +1,83 @@
+
+        int currentOffset;
+        // Need to use TreeMap here, so that we maintain the order in which each value (with offset) gets inserted,
+        // (which is in the same order was document gets parsed) so we store offsets in right order.
```
Nit: s/was document/the document/

kkrik-es commented on server/src/main/java/org/elasticsearch/index/mapper/FieldArrayContext.java:
```diff
@@ -0,0 +1,83 @@
+                for (int ord : offsetToOrd) {
+                    streamOutput.writeVInt(BitUtil.zigZagEncode(ord));
+                }
+                context.doc().add(new BinaryDocValuesField(fieldName, streamOutput.bytes().toBytesRef()));
```
I think you can skip the context dep here, just have this function return `BinaryDocValuesField`.

martijnvg replied:
This function will need to return a list/array of `BinaryDocValuesField` entries, I think it is better to just add the `BinaryDocValuesField` entries to the document, instead of creating another list/array that will be used to add fields to the document.

kkrik-es commented on server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java:
```diff
@@ -411,13 +417,34 @@ public KeywordFieldMapper build(MapperBuilderContext context) {
+
+            var sourceKeepMode = this.sourceKeepMode.orElse(indexSourceKeepMode);
+            String offsetsFieldName;
+            if (context.isSourceSynthetic()
```
Use `canAddIgnoredField` instead.

martijnvg replied:
I think that isn't available in `MapperBuilderContext?`

kkrik-es commented on server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java:
```diff
@@ -1066,17 +1097,42 @@ private KeywordFieldMapper(
+        }
+
+        boolean indexed = indexValue(context, value);
+        if (offsetsFieldName != null && context.isImmediateParentAnArray() && context.getRecordedSource() == false) {
```
Same, use `canAddIgnoredField` instead of `getRecordedSource`.

kkrik-es replied:
Not updated?

lkts replied:
+1

kkrik-es commented on server/src/main/java/org/elasticsearch/index/mapper/Mapper.java:
```diff
@@ -212,4 +212,19 @@ public static FieldType freezeAndDeduplicateFieldType(FieldType fieldType) {
     public abstract int getTotalFieldsCount();
+
+    /**
+     * @return whether this mapper support storing leaf array elements natively when synthetic source is enabled.
```
Nit: s/support/supports/

kkrik-es commented on server/src/main/java/org/elasticsearch/index/mapper/Mapper.java:
```diff
@@ -212,4 +212,19 @@ public static FieldType freezeAndDeduplicateFieldType(FieldType fieldType) {
+    }
+
+    /**
+     * @return the offset field name use the store offsets iff {@link #supportStoringArrayOffsets(DocumentParserContext)} returns
```
This also needs a slight rewording.

kkrik-es commented on server/src/test/java/org/elasticsearch/index/mapper/KeywordFieldTypeTests.java:
```diff
@@ -244,7 +244,8 @@ public void testFetchSourceValue() throws IOException {
             Integer.MAX_VALUE,
-            IndexVersion.current()
+            IndexVersion.current(),
+            randomFrom(Mapper.SourceKeepMode.values())
```
Let's add a few more tests:
 - Leaf arrays containing leaf arrays
 - Object arrays containing leaf arrays
 - No offset tracking when synthetic source is disabled
 - No offset tracking when a parent is stored in ignored source

martijnvg replied:
These tests exists now in `OffsetDocValuesLoaderTests` and `SyntheticSourceNativeArrayIntegrationTests`.

kkrik-es commented on server/src/main/java/org/elasticsearch/index/mapper/SortedSetWithOffsetsDocValuesSyntheticFieldLoaderLayer.java:
```diff
@@ -0,0 +1,175 @@
+        BinaryDocValues offsetDocValues = DocValues.getBinary(leafReader, offsetsFieldName);
+
+        ImmediateDocValuesLoader loader = new ImmediateDocValuesLoader(valueDocValues, offsetDocValues);
+        docValues = loader;
```
Nit: skip intermediate variable `loader`.

kkrik-es commented on server/src/main/java/org/elasticsearch/index/mapper/SortedSetWithOffsetsDocValuesSyntheticFieldLoaderLayer.java:
```diff
@@ -0,0 +1,175 @@
+        }
+    }
+
+    static final class ImmediateDocValuesLoader implements DocValuesLoader {
```
Maybe `CompoundDocValuesLoader`? Not sure what `Immediate` means here.

kkrik-es commented on server/src/main/java/org/elasticsearch/index/mapper/SortedSetWithOffsetsDocValuesSyntheticFieldLoaderLayer.java:
```diff
@@ -0,0 +1,175 @@
+
+                    long ord = ords[offset];
+                    BytesRef c = valueDocValues.lookupOrd(ord);
+                    b.utf8Value(c.bytes, c.offset, c.length);
```
This is keyword-specific, correct? Maybe add a comment for it.

kkrik-es reviewed (commented):
Looks good, a few nits and testing comments.

lkts commented on server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java:
```diff
@@ -688,11 +695,12 @@ private static void parseNonDynamicArray(
         // Check if we need to record the array source. This only applies to synthetic source.
         boolean canRemoveSingleLeafElement = false;
-        if (context.canAddIgnoredField()) {
+        if (context.canAddIgnoredField() && (parsesArrayValue(mapper) == false && supportStoringArrayOffsets == false)) {
```
I think i spend 10+ minutes trying to understand this line. Let's at least comment what's going on. I wonder if we can structure this better. Can we maybe split parsing logic based on `parsesArrayValue` into different methods?

lkts replied:
Actually it looks like if `parsesArrayValue(mapper) == true` we won't even get here?

lkts commented on server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java:
```diff
@@ -411,13 +417,34 @@ public KeywordFieldMapper build(MapperBuilderContext context) {
+
+            var sourceKeepMode = this.sourceKeepMode.orElse(indexSourceKeepMode);
+            String offsetsFieldName;
+            if (context.isSourceSynthetic()
```
Don't we need to check that `doc_values` are enabled?

lkts replied:
I really don't like how this check repeats a very similar one in `DocumentParser` that decides to apply ignored source.

martijnvg replied:
Good point, we need to check whether doc value is enabled. I updated this check and added tests.

lkts commented on server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java:
```diff
@@ -729,13 +737,16 @@ private static void parseNonDynamicArray(
-        // elements from regular source loading that are then skipped from the synthesized array source.
-        // To prevent this, we track that parsing sub-context is within array scope.
-        context = context.maybeCloneForArray(mapper);
+        if (parsesArrayValue(mapper) == false && supportStoringArrayOffsets == false) {
```
`maybeCloneForArray` only cares about objects, is this check relevant?

martijnvg replied:
Yes, turns out this wasn't needed at all.

martijnvg replied:
It was there, because of how this PR worked initially, but currently not needed anymore.

lkts reviewed (commented):
I've asked some questions. 

I am also not too happy with added complexity in `DocumentParser` and especially the split of the decision to use or not to use ignored source between `DocumentParser` and `KeywordFieldMapper`. I wonder if we can avoid that.

martijnvg committed:
Merge remote-tracking branch 'es/main' into synthetic_source_encode_arrays

martijnvg committed:
remove `parsesArrayValue(mapper) == false` checks

martijnvg committed:
cleanup DocumentParser and
add tests with array in parent object field.

martijnvg committed:
ensure doc values enabled and added more tests

kkrik-es commented on server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java:
```diff
@@ -615,8 +613,12 @@ private static void throwOnCreateDynamicNestedViaCopyTo(Mapper dynamicObjectMapp
     }
 
     private static void parseArray(DocumentParserContext context, String lastFieldName) throws IOException {
+        // Record previous immediate parent, so that it can be reset after array has been parsed:
```
Nit: s/:/./

martijnvg commented on server/src/main/java/org/elasticsearch/index/mapper/FieldArrayContext.java:
```diff
@@ -0,0 +1,83 @@
+        arrayOffsets.nullValueOffsets.add(nextOffset);
+    }
+
+    void maybeRecordEmptyArray(String field) {
```
This is invoked when END_ARRAY is encountered. At this time, there may be an entry for the `field`, if not an empty `Offsets` will be added, that later will be used the record an empty error in the related offset set.

martijnvg committed:
Merge remote-tracking branch 'es/main' into synthetic_source_encode_arrays

martijnvg committed:
iter

martijnvg committed:
s/:/./

kkrik-es commented on server/src/main/java/org/elasticsearch/index/mapper/Mapper.java:
```diff
@@ -214,14 +214,14 @@ public static FieldType freezeAndDeduplicateFieldType(FieldType fieldType) {
 
     /**
-     * @return the offset field name use the store offsets iff {@link #supportStoringArrayOffsets(DocumentParserContext)} returns
+     * @return the offset field name used the store offsets iff {@link #supportStoringArrayOffsets(DocumentParserContext)} returns
```
s/used the/used to/ ?

martijnvg committed:
iter

martijnvg committed:
Merge remote-tracking branch 'es/main' into synthetic_source_encode_arrays

kkrik-es commented on server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java:
```diff
@@ -611,6 +613,12 @@ private static void throwOnCreateDynamicNestedViaCopyTo(Mapper dynamicObjectMapp
     private static void parseArray(DocumentParserContext context, String lastFieldName) throws IOException {
+        // Record previous immediate parent, so that it can be reset after array has been parsed.
+        // This is for recording array offset with synthetic source. Only if the immediate parent is an array,
+        // then we offsets can be accounted accurately.
```
Nit: s/we offsets/the offsets/

kkrik-es commented on server/src/main/java/org/elasticsearch/index/mapper/FieldArrayContext.java:
```diff
@@ -0,0 +1,83 @@
+
+        int currentOffset;
+        // Need to use TreeMap here, so that we maintain the order in which each value (with offset) gets inserted,
+        // (which is in the same order the document gets parsed) so we store offsets in right order.
```
Don't we need to use the TreeMap here to reflect the sorted order for retrieving entries from SortedSetDocValues?

martijnvg replied:
Yes, the ordering of `TreeMap` matches the ordering in `SortedSetDocValues`.

kkrik-es commented on server/src/main/java/org/elasticsearch/index/mapper/SortedSetWithOffsetsDocValuesSyntheticFieldLoaderLayer.java:
```diff
@@ -0,0 +1,174 @@
+    }
+
+    static int[] parseOffsetArray(StreamInput in) throws IOException {
+        int[] offsetToOrd = new int[BitUtil.zigZagDecode(in.readVInt())];
```
Nit: the use of `zigzag` feels somewhat arbitrary here. Maybe move this helper function to `FieldArrayContext` where `zigzag` is also used for storing values.

martijnvg committed:
Store offsets in sorted doc values field instead of binary doc values field

martijnvg committed:
applied feedback

martijnvg commented:
With the last two commits the disk usage for the offset fields is much lower:

```
 field name (from different shard)                                                   | Size     | Relative to total shard size 
apache.access.remote_addresses.offsets                                               | 3.0b     | 0.00%
event.category.offsets                                                               | 3.0b     | 0.00%
host.mac.offsets                                                                     | 3.0b     | 0.00%
event.category.offsets                                                               | 3.0b     | 0.00%
event.type.offsets                                                                   | 3.0b     | 0.00%
host.mac.offsets                                                                     | 3.0b     | 0.00%
tags.offsets                                                                         | 5.1mb    | 0.03%
log.flags.offsets                                                                    | 4.0mb    | 0.02%
event.category.offsets                                                               | 94.4kb   | 0.00%
event.type.offsets                                                                   | 94.4kb   | 0.00%
host.ip.offsets                                                                      | 30.9kb   | 0.00%
host.mac.offsets                                                                     | 30.9kb   | 0.00%
tags.offsets                                                                         | 2.5mb    | 0.03%
log.flags.offsets                                                                    | 1.7mb    | 0.02%
event.category.offsets                                                               | 41.1kb   | 0.00%
event.type.offsets                                                                   | 41.1kb   | 0.00%
host.ip.offsets                                                                      | 17.9kb   | 0.00%
host.mac.offsets                                                                     | 17.9kb   | 0.00%
log.flags.offsets                                                                    | 1.7mb    | 0.26%
event.type.offsets                                                                   | 17.1kb   | 0.00%
host.mac.offsets                                                                     | 3.0b     | 0.00%
log.flags.offsets                                                                    | 1017.0b  | 0.00%
event.category.offsets                                                               | 3.0b     | 0.00%
event.type.offsets                                                                   | 3.0b     | 0.00%
host.mac.offsets                                                                     | 3.0b     | 0.00%
event.category.offsets                                                               | 3.0b     | 0.00%
event.type.offsets                                                                   | 3.0b     | 0.00%
host.mac.offsets                                                                     | 3.0b     | 0.00%
log.flags.offsets                                                                    | 3.0b     | 0.00%
tags.offsets                                                                         | 29.1kb   | 0.00%
event.category.offsets                                                               | 3.0b     | 0.00%
event.type.offsets                                                                   | 3.0b     | 0.00%
host.mac.offsets                                                                     | 3.0b     | 0.00%
nginx.access.remote_ip_list.offsets                                                  | 3.0b     | 0.00%
related.user.offsets                                                                 | 3.0b     | 0.00%
tags.offsets                                                                         | 25.7kb   | 0.01%
event.category.offsets                                                               | 3.0b     | 0.00%
event.type.offsets                                                                   | 3.0b     | 0.00%
host.mac.offsets                                                                     | 3.0b     | 0.00%
log.flags.offsets                                                                    | 24.2kb   | 0.01%
event.category.offsets                                                               | 17.2kb   | 0.01%
event.type.offsets                                                                   | 17.2kb   | 0.01%
related.user.offsets                                                                 | 17.2kb   | 0.01%
host.mac.offsets                                                                     | 3.0b     | 0.00%
host.mac.offsets                                                                     | 3.0b     | 0.00%
redis.slowlog.args.offsets                                                           | 24.5mb   | 0.91%
host.mac.offsets                                                                     | 3.0b     | 0.00%
related.user.offsets                                                                 | 3.8mb    | 0.43%
event.category.offsets                                                               | 2.3mb    | 0.26%
event.type.offsets                                                                   | 1.1mb    | 0.13%
host.mac.offsets                                                                     | 435.7kb  | 0.05%
tags.offsets                                                                         | 93.7kb   | 0.01%
related.hosts.offsets                                                                | 3.0b     | 0.00%
host.mac.offsets                                                                     | 376.3kb  | 0.02%
tags.offsets                                                                         | 22.9kb   | 0.00%
```

lkts commented on test/framework/src/main/java/org/elasticsearch/index/mapper/MapperTestCase.java:
```diff
@@ -1707,7 +1707,7 @@ public void testSyntheticSourceKeepArrays() throws IOException {
         DocumentMapper mapperAll = createSytheticSourceMapperService(mapping(b -> {
             b.startObject("field");
-            b.field("synthetic_source_keep", randomFrom("arrays", "all"));  // Both options keep array source.
+            b.field("synthetic_source_keep", randomFrom("all"));  // Only option all keeps array source.
```
Should this change only be done for keyword field?

kkrik-es replied:
This is not needed even for keyword? All values should work.

martijnvg replied:
I think we can make this change only for keyword field mapper. Given that other field types still fallback to ignored source if source keep is set to arrays.

martijnvg replied:
I pushed: bbee16082e99115b7ec14fb857781fefe191b0db

lkts reviewed (approved):
Thank you for iterating on this.

martijnvg committed:
Merge remote-tracking branch 'es/main' into synthetic_source_encode_arrays

martijnvg committed:
iter testSyntheticSourceKeepArrays() test

martijnvg committed:
add index version check

martijnvg committed:
iter test

Commit:
[CI] Auto commit changes from spotless

martijnvg committed:
Merge remote-tracking branch 'es/main' into synthetic_source_encode_arrays

martijnvg committed:
Merge remote-tracking branch 'es/main' into synthetic_source_encode_arrays

martijnvg committed:
cleanup supportStoringArrayOffsets(...) method

martijnvg committed:
renamed test suites

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  8.x  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 113757`
